module.c

```diff
@@ -442,6 +514,78 @@ static PyObject *context_gc(ContextData *self) {
 	Py_RETURN_NONE;
 }
 
+static JSValue js_c_function(
+    JSContext *ctx, JSValueConst this_val, int argc, JSValueConst *argv, int magic) {
+	ContextData *self = (ContextData *)JS_GetContextOpaque(ctx);
+	PythonCallableNode *node = self->python_callables;
+	while (node && node->magic != magic) {
+		node = node->next;
+	}
+	if (!node) {
+		return JS_ThrowInternalError(self->context, "Internal error.");
+	}
+
+	PyObject *args = PyTuple_New(argc);
+	if (!args) {
+		return JS_ThrowOutOfMemory(self->context);
+	}
+	for (int i = 0; i < argc; ++i) {
+		PyObject *arg = quickjs_to_python(self, JS_DupValue(self->context, argv[i]));
+		PyTuple_SET_ITEM(args, i, arg);
+	}
+
+	PyObject *result = PyObject_CallObject(node->obj, args);
+	Py_DECREF(args);
+	if (!result) {
+		return JS_ThrowInternalError(self->context, "Python call failed.");
+	}
+	JSValue js_result = JS_NULL;
+	if (python_to_quickjs_possible(self, result)) {
+		js_result = python_to_quickjs(self, result);
+	}
+	Py_DECREF(result);
+	return js_result;
+}
+
+static PyObject *context_add_callable(ContextData *self, PyObject *args) {
+	const char *name;
+	PyObject *callable;
+	if (!PyArg_ParseTuple(args, "sO", &name, &callable)) {
+		return NULL;
+	}
+	if (!PyCallable_Check(callable)) {
+		PyErr_SetString(PyExc_TypeError, "Argument must be callable.");
+		return NULL;
+	}
+
+	PythonCallableNode *node = PyMem_Malloc(sizeof(PythonCallableNode));
+	if (!node) {
+		return NULL;
+	}
+	Py_INCREF(callable);
+	self->can_release_gil = 0;
+	node->magic = 0;
+	if (self->python_callables) {
+		node->magic = self->python_callables->magic + 1;
+	}
+	node->obj = callable;
+	node->next = self->python_callables;
+	self->python_callables = node;
+
+	JSValue function = JS_NewCFunctionMagic(
+	    self->context,
+	    js_c_function,
+	    "python",  // TODO: Name of the function should not matter that much?
+	    1,  // TODO: Should we allow setting the .length of the function to something other than 1?
+	    JS_CFUNC_generic_magic,
+	    node->magic);
+	JSValue global = JS_GetGlobalObject(self->context);
+	// If this fails we don't notify the caller of this function.
+	JS_SetPropertyStr(self->context, global, name, function);
+	JS_FreeValue(self->context, global);
```

Comment:
```suggestion

	JSContext *ctx = self->context;
	JSValue function = JS_NewCFunctionMagic(
	    ctx,
	    js_c_function,
	    "python",  // TODO: Name of the function should not matter that much?
	    1,  // TODO: Should we allow setting the .length of the function to something other than 1?
	    JS_CFUNC_generic_magic,
	    node->magic);

	if (!prepare_call_js(self)) {
		return NULL;
	}
	char setfunc[strlen(name) + 17];
	sprintf(setfunc, "(func => %s = func)", name);
	JSValue ret, ret2;
	ret = JS_Eval(ctx, setfunc, strlen(setfunc), "<input>", JS_EVAL_TYPE_GLOBAL);

	JSValueConst *jsargs = malloc(sizeof(JSValueConst));
	jsargs[0] = function;
	ret2 = JS_Call(ctx, ret, JS_UNDEFINED, 1, jsargs);
	JS_FreeValue(ctx, ret);
	free(jsargs);

	end_call_js(self);
	JS_FreeValue(ctx, function);
	if (!JS_IsException(ret2)) {
		JS_FreeValue(ctx, ret2);
		self->python_callables = node;
		return Py_True;
	} else {
		JS_FreeValue(ctx, ret2);
		return Py_False;
	}
```
I try to add callable via `JS_Call`, its returned status is right, but another error occurred.
```
ERROR: test_add_callable (test_quickjs.FunctionTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\quickjs-m\test_quickjs.py", line 434, in test_add_callable
    self.assertEqual(f(), 42)
  File "D:\quickjs-m\quickjs\__init__.py", line 46, in __call__
    return future.result()
  File "concurrent\futures\_base.py", line 432, in result
  File "concurrent\futures\_base.py", line 388, in __get_result
  File "concurrent\futures\thread.py", line 57, in run
  File "D:\quickjs-m\quickjs\__init__.py", line 91, in _call
    result = self._f(*[convert_arg(a) for a in args])
_quickjs.JSException: ReferenceError: pfunc is not defined
    at f (<input>)
```